### ISSUE TITLE
Proposal: Enable opt-in HTTPS between router and app containers

### ISSUE THREAD

**Issue by krancour** [deis/deis#4629]
See issue #4611 

Deis currently uses "mullet TLS--" security in the front, party in the back.  Having come from healthcare, I understand all too well how this rules out the possibility of using Deis to host any workloads that deal with PHI.  This is a gap that needs to be closed.

This issue isn't meant to duplicate #4611, but rather to serve as a forum for discussing proposed designs that address this.  I'll start by throwing my proposal into the ring.  Feedback is welcome, as are alternate proposals.

Approach
-------------

I think we can pretty easily go 80/20 on this and get ourselves most of the way to where we need to be pretty quickly and then improve incrementally from there going forward.  Initially, I'd propose that we support the the following, at minimum:

* 100% opt-in on an application basis so as not to impact existing, deployed applications upon platform upgrade.
* _Option_ for end-to-end HTTPS needs to be there.  If when selecting this option, it proves infeasible to _close_ the HTTP route into the same application in the first iteration, so be it.
* Self-signed certificates are good enough, initially.  It's only the router that needs to trust the certs-- not the end user's agent (since Deis can already terminate TLS at the router using a proper cert provided by the operator, signed by a known and trusted CA).

Possible implementation
---------------------------------

I believe operators should opt-in to this on an application basis by setting an environment variable via `deis config:set`.

The following is all WRT to buildpack-based applications:

One possible implementation is to generate a self-signed certificate at build-time.  I imagine the easiest way of doing this being modifications to the slugbuilder such that after the buildpack(s) have generated the application's deployable artifacts, the generated cert and key are added to those right before compressing those artifacts into a slug (tarball).  A corresponding change to the slugrunner can start Nginx as a proxy in front of the application to terminate TLS.

Open questions and concerns
----------------------------------------

I think experimentation is the best way to gain insight into the following:

* Will application containers _all_ include a self-signed cert and Nginx?  Or just those that have opted into it?  This depends on the availability of relevant information (environment variable) at build time.  Fortunately, the environment variables for each application are _also_ inserted into etcd, so we probably do have access to this information at build-time.

* If one opts into end-to-end HTTPS, can we expose the secure port in the application containers _instead of_ the insecure port?  Or, initially, do we expose it _in addition to_ the insecure port?  Proposing we shoot for the former; settle for the latter if necessary (for the first iteration).

* How are routing and publishing affected?  Publishing may or may not be affected depending on the previous bullet.  Routing rules will be slightly different for anyone any app that opts into this.

* How does this affect health checks and liveliness checks?

* How does it work with k8s as the scheduler?  With k8s as the scheduler, publisher isn't really relevant.  Router proxies service URLs.  Those service URLs likely need to change for any app that has opted into this.

* How does this work for developers pushing their own Dockerfile or for those using `deis pull`?  My personal take is that anyone who's already exercising that level of control over their images is responsible for exposing the secure endpoint themselves.  We would just need to clearly document how the router plugs into that in such scenarios.

**Comment by krancour** [comment]
> Will application containers all include a self-signed cert and Nginx? Or just those that have opted into it? This depends on the availability of relevant information (environment variable) at build time.

See #4569 for a clear understanding of how this works:

> The builder produces images that are devoid of "last mile" configs. The builder triggers a release by calling a hook in the controller, and the controller does bake configs into an image during the release process.

> The process looks something like this:

> build image + config == release image

If an HTTPS application endpoint is enabled or disabled by an environment variable set through `deis config:set`, that config gets baked into a new release image, but the builder is not at all involved in that process.

This means when the builder produces a "build image" that is absent all that config, it _must_ include a self-signed cert and Nginx at that time because it is unknown whether or not they will be needed down the road.

At runtime, however, we _can_ eliminate some overhead by not starting Nginx when the app starts if the configuration indicates it's not needed.

List of changes for my proposed approach seems to be shaping up as follows:

* Modify slug-builder to generate self-signed cert
* Modify slug-runner Dockerfile to expose secure port
* Modify slug-runner to include Nginx, including some static configuration
* Modify slug-runner so app start script also starts Nginx (or doesn't if not needed)
* Modify router confd template for router config to use HTTPS for back-end proxy target as appropriate

**Comment by krancour** [comment]
@technosophos and @mboersma would love your insight on the short list above.  Am I missing anything obvious?

**Comment by sbuss** [comment]
We've been brainstorming a solution that would require no changes from users, though it does require two separate HTTPS connections (but I think TLS session caching would work). It should also be trivially portable to a kubernetes pod-style deployment.

TL;DR: Run an nginx instance on every machine in the cluster, which terminates a TLS connection from the router (not from the client), and reverse proxies on localhost to the correct container. This works for buildpack/Dockerfile/`deis pull` deploys.

## Using the core (non-k8s/mesos) infrastructure:

I'll build up from the app to the `deis-router`:

* The app you're deploying does not need to be altered.
* The `publisher` should continue publishing `hostAndPort` to `/deis/services/<app-name>`.
* Schedule a global fleet unit (call it `deis-local-router`) which runs nginx on every CoreOS machine and behaves a lot like the current `deis-router`, except:
    * Filter entries in `/deis/services/<app-name>` where `host:port` matches the hosts's IP
    * All upstreams are only on localhost
    * Only accept traffic on port 443 using a deis-internal TLS cert
* The `deis-router` keeps the same configuration except for ignoring the `:port` part of the `host:port` entry, and instead always upstreams to `host:443`, where `deis-local-router` is listening

The request path would then look like this:

```
client --(TLS w/ external cert)--> deis-router --(TLS w/ deis-internal cert)-+
                                                                             |
                              container:port  <--(localhost)-- host1:443  <--+
```

## Using kubernetes

This `deis-local-router` pattern should easily translate to an `nginx` container running alongside an app container in a `k8s pod`. I have not yet explored what might need to change to support this approach, though.

## Addressing open questions

I think this approach answers all of the open questions:

> Will application containers all include a self-signed cert and Nginx? Or just those that have opted into it?

The application containers won't need to worry at all about the cert or nginx, since a host-local nginx instance will handle that for them. All traffic received by the container will be unencrypted since it will be confined to localhost.

> If one opts into end-to-end HTTPS, can we expose the secure port in the application containers instead of the insecure port? Or, initially, do we expose it in addition to the insecure port? Proposing we shoot for the former; settle for the latter if necessary (for the first iteration).

Since the apps don't need to have knowledge of the backend security, always use the encrypted connection pattern that `deis-router-local` enables.

> How are routing and publishing affected? Publishing may or may not be affected depending on the previous bullet. Routing rules will be slightly different for anyone any app that opts into this.

Routing is affected by `deis-router` terminating public-facing TLS and starting a new TLS connection to the backend. `deis-router` would not route directly to the app's port on the target host, but instead directly to `host:443` which `deis-router-local` is listening to.

Publishing is not affected.

> How does this affect health checks and liveliness checks?

If this is a question about rolling/no-downtime deploys, these checks would all be confined to localhost in `deis-router-local`.

> How does it work with k8s as the scheduler? With k8s as the scheduler, publisher isn't really relevant. Router proxies service URLs. Those service URLs likely need to change for any app that has opted into this.

Pods deployed with a `deis-router-local` instance solve this well.

> How does this work for developers pushing their own Dockerfile or for those using deis pull?

Nothing needs to change because the TLS termination will still happen outside of the app container.

**Comment by krancour** [comment]
@sbuss, that is a great proposal!  I think I'd prefer your approach.  Nothing problematic is jumping out at me.  I'll play with this idea tomorrow... of course feel free to PR this if you have it working.  :wink:

**Comment by sbuss** [comment]
Glad you like it! Also thanks to @myyk for developing the proposal with me today.

> of course feel free to PR this if you have it working.  :wink:

Alas, we haven't written any code for it yet, but will hopefully have time this week.

**Comment by krancour** [comment]
@sbuss, the one problem I am potentially seeing now is that this may not work with flannel's overlay network in play.  With flannel, the IPs that appear in /deis/services are not host IPs; they are IPs for a container.  Figuring out the corresponding host IP would be the trick.  Ultimately, that may mean publisher changes are still called for.

**Comment by krancour** [comment]
Basically, what we'd need in the router is that for every app, the router would have to proxy traffic not to the hosts of each app, but to the flannel-assigned overlay IP for the "deis-router-host" containers, but ONLY such containers that reside on the same host as at least one instance of the given application.  This sounds very tricky, and even if it works, it could lead to some very unbalanced traffic.  The routet would be round robining requests to the local routers, but some local routers maybve proxying more or fewer instances of the given app than are its peers.

Just some things to think through.

**Comment by krancour** [comment]
As it turns out, weights can be assigned to upstream servers in Nginx.  This solves the problem of unbalanced traffic to the app containers.

**Comment by krancour** [comment]
And publisher still publishes the IP of a container's host; not the overlay IP.  So that's another piece of the puzzle solved (for schedulers where publisher is used; k8s still to be solved, but the sidecar container @sbuss mentioned would probably work).

**Comment by krancour** [comment]
Though it looks as if it may be tricky, I think potentially, the publisher can be modified to maintain a map of apps to instance counts.

This can be published to etcd like so: `/deis/secureServices/<app>/<host ip> = <count>`

The router can use those values as the weights for upstreams servers-- which are local routers.

**Comment by jchauncey** [comment]
+1 on  @sbuss 's idea. Thats basically what we did at rally to paas our compliance tests for the paas

**Comment by krancour** [comment]
@sbuss' idea is the front-runner.  @helgi and I have spent the day exploring various aspects of it.  It doesn't look easy, but it looks possible.  WIP PR to follow shortly.

**Comment by krancour** [comment]
Capturing a few other thoughts from a discussion @helgi and I had-- just so they're not forgotten and to give the community a chance to sound off.

We've been working toward proving the feasibility of @sbuss' proposed design and things look good _so far_.

One thing we have been discussing is how operators will opt-in to this either on an app by app basis or across the board.  My recent line of thinking had been that anyone who secures the connection from the browser to the router (Deis already permits this), would automatically get the hop from the router to the "local router" secured as well.  This has the advantage of improved security being a _defualt_, essentially, but @helgi has pointed out the overhead of terminating and re-initiating TLS.

Something we agreed we could play with is making the existing router a pass-through in such cases and not terminate TLS at all until the request hits the local routers.  Local routers would use the very same certs for this that the edge router would have used previously.  These are _not_ self-signed certs.  They are the actual certs (presumably from a trusted CA) that operators can already manage today through established `deis` CLI workflows.

If this works, folks who have already opted into HTTPS for the hop from the browser to the router would now get HTTPS end-to-end with virtually no performance penalty and no self-signed certs in play.  It also, probably, has the advantage of being simple to implement.

Opening this up for further feedback, including criticism and alternate suggestions...

**Comment by sbuss** [comment]
> With flannel, the IPs that appear in /deis/services are not host IPs; they are IPs for a container. 

Oh, I assumed this would be the IP of the pod

> My recent line of thinking had been that anyone who secures the connection from the browser to the router (Deis already permits this), would automatically get the hop from the router to the "local router" secured as well. This has the advantage of improved security being a default, essentially

+1

> but @helgi has pointed out the overhead of terminating and re-initiating TLS may outweigh the advantages for some operators.

Isn't the overhead for this fairly small, though? You can make the `deis-router`s reuse the TLS session parameters via `proxy_ssl_session_reuse  on`. Reusing the sessions should make the rest of it not too inefficient.

> Something we agreed we could play with is making the existing router a pass-through in such cases and not terminate TLS at all until the request hits the local routers. Local routers would use the very same certs for this that the edge router would have used previously. These are not self-signed certs. They are the actual certs (presumably from a trusted CA) that operators can already manage today through established deis CLI workflows.

This would be wonderful and is definitely my preferred solution, but we were unsure how the TLS session negotiation would happen. If the request goes `client -> deis-router -> deis-local-router -> container`, then wouldn't the roundtrips to establish the TLS session possibly end up at different `deis-local-router`s?

**Comment by krancour** [comment]
> then wouldn't the roundtrips to establish the TLS session possibly end up at different deis-local-routers

Already we run multiple edge routers (load balanced by an ELB, for instance) and that doesn't happen.  Based on that alone, I would think that shouldn't happen when the TLS termination gets kicked down the road to the local routers.

**Comment by jeff-lee** [comment]
It looks like we're going to have to run a custom router in order to support a larger large_client_header_buffers setting.  Would this change mean we would need custom local routers as well if we wanted to enable ssl to the apps?

**Comment by myyk** [comment]
> Already we run multiple edge routers (load balanced by an ELB, for instance) and that doesn't happen. Based on that alone, I would think that shouldn't happen when the TLS termination gets kicked down the road to the local routers.

@krancour, under the existing architecture the deis routers can terminate traffic that has been routed from an multiple edge routers that are either recreating an encrypted connection or by doing OSI L4 load balancing of the TCP connection. To route TCP traffic at the layer 4 from the deis routers to the local routers, we would need to make sure that the deis routers don't need to do any http mangling (L7 load balancing) like they are doing now. It seems like that functionality can be pushed to the local routers.

> It looks like we're going to have to run a custom router in order to support a larger large_client_header_buffers setting. Would this change mean we would need custom local routers as well if we wanted to enable ssl to the apps?

I'd imagine if all the L7 load balancing is moving to the local routers, then you'd only need to customize the local routers.

**Comment by krancour** [comment]
Much of the strategy that's been discussed above has been only within the context of fleet as the scheduler.  Because routing for apps deployed in k8s works a bit differently, supporting end-to-end HTTPS for the k8s "tech preview" requires a different strategy.  After much discussion, we've decided that that's the only bit of any effort that's spent here that's likely to carry forward into v2, and as such, we're going to focus out efforts on solving this problem _only_ for k8s.

**Comment by krancour** [comment]
This will be implemented in v2 and is tracked in deis/workflow#355.  As such, I'm closing this issue.
